src/library/scala/util/Using.scala

```diff
@@ -0,0 +1,286 @@
+package scala.util.control
+
+import java.util.concurrent.atomic.AtomicBoolean
+
+import scala.util.Try
+
+/** A utility for performing automatic resource management. It can be used to perform an
+  * operation using resources, after which it will close the resources, in reverse order
+  * of their creation. The resource opening, operation, and resource closing are wrapped
+  * in a `Try`.
+  *
+  * If more than one exception is thrown by the operation and closing resources,
+  * the exception thrown ''first'' is returned within the `Try`, with the other exceptions
+  * [[java.lang.Throwable.addSuppressed(Throwable) added as suppressed exceptions]]
+  * to the one thrown first. This is the case ''unless'' a later exception is
+  * [[scala.util.control.NonFatal fatal]], and the one preceding it is not. In that case,
+  * the first exception is added as a suppressed exception to the fatal one, and the fatal
+  * one is thrown. If an exception is a
+  * [[scala.util.control.ControlThrowable ControlThrowable]], no exception will be added to
+  * it as a suppressed exception.
+  *
+  * @example
+  * {{{
+  * val lines: Try[List[String]] = Using(resource1) { r1 =>
+  *   r1.lines.toList
+  * }
+  * }}}
+  *
+  * @example
+  * {{{
+  * val lines: Try[Seq[String]] = for {
+  *   r1 <- Using(resource1)
+  *   r2 <- Using(resource2)
+  *   r3 <- Using(resource3)
+  *   r4 <- Using(resource4)
+  * } yield {
+  *   // use your resources here
+  *   r1.lines ++ r2.lines ++ r3.lines ++ r4.lines
+  * }
+  * }}}
+  */
+final class Using[R](resource: => R) {
+  private[this] val used = new AtomicBoolean(false)
+
+  /** Performs an operation using a resource, and then closes the resource,
+    * even if the operation throws an exception.
+    *
+    * @param f the operation to perform
+    * @param r an implicit [[Using.Resource]]
+    * @tparam A the return type of the operation
+    * @throws java.lang.IllegalStateException if the resource has already been used
```

Comment:
I am open to being convinced that the resource can be generated again using the by-name parameter, and that the `Using` instance can be used arbitrarily many times.

Reply:
@NthPortal  Sorry. I forgot to review.

Reply:
@NthPortal It looks good to me as far as I can see